doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,968 @@
+    * *cmake*: cmake definitions to check the python environment. Required to generate our passes test suites.
+    * *doc*: contains the sources of this tutorial, in case you find a shaming typo.
+    * *llvm-passes*: contains one subdirectory per pass, and a ``CMakeList.txt`` used to generate the passes.
+    * *tests*: tests and validation for our passes, contains one directory per pass. The tests are using llvm-lit, the LLVM integrated validation tool.
+    * *CMakeList.txt*: the file used to generate the required Makefiles
+
+
+Let's obfuscate!
+================
+
+Now that the environment is ready we will start writing our obfuscating pass. You may have noticed that there already is an ``ObfuscateZero`` dir in ``$PASSDIR/llvm-passes``.
+This is the pass we are going to reproduced step by step. So unless you want to get spoiled don't look at it yet.
+
+Now we have to deal with the hardest part of LLVM pass development (and software development in general), namely finding a name for our project.
+Since I am not really inspired and ``ObfuscateZero`` is already taken, let's call our new pass *MyPass*.
+
+We need a new directory for our pass:
+    ``>$ mkdir $PASSDIR/llvm-passes/MyPass``
+
+And we will write the pass in ``$PASSDIR/llvm-passes/MyPass/MyPass.cpp``.
+
+
+One with nothing
+****************
+
+The minimal compiling code for an LLVM pass is the following. It is explained `there <http://llvm.org/docs/WritingAnLLVMPass.html#basic-code-required>`_ so I won't explain it again and focus on the obfuscation part.
+
+.. code:: C++
+
+    #include "llvm/Pass.h"
+    #include "llvm/IR/Function.h"
+    #include "llvm/Support/raw_ostream.h"
+
+    #include "llvm/IR/LegacyPassManager.h"
+    #include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+    using namespace llvm;
+
+    namespace {
+    class MyPass : public BasicBlockPass {
+    public:
+      static char ID;
+
+      MyPass() : BasicBlockPass(ID) {}
+
+      virtual bool runOnBasicBlock(BasicBlock &BB) override {
+        errs() << "I m running on a block...\n";
+        return false;
+      }
+
+    };
+    }
+
+    char MyPass::ID = 0;
+    static RegisterPass<MyPass> X("MyPass", "Obfuscates zeroes",
+                                         false, false);
+
+    // register pass for clang use
+    static void registerMyPassPass(const PassManagerBuilder &,
+                                   PassManagerBase &PM) {
+      PM.add(new MyPass());
+    }
+    static RegisterStandardPasses
+        RegisterMBAPass(PassManagerBuilder::EP_EarlyAsPossible,
+                        registerMyPassPass);
+
+If you have been paying attention so far you should remember that we are going to obfuscate null literals.
+And to do so we will randomly pick two variables reachable from where the replacement occurs.
+So, in order to keep the pass as simple as possible we are going to work at the basic bloc level, this way there will be no reachability problems with the variables we encounter.
```

Comment:
even better, generating different patterns in multiple basic blocks and using phi to select, or even part of the pattern in various block and recomposing it in various paths. It might increase the entropy?